DRIVERS-3588: Add Claude Code skill for retiring EOL server version references (#1970)
mongodb/specifications: .claude/skills/retiring-server-versions/SKILL.md

---
name: retiring-server-versions
description: >-
  Use when raising the minimum supported MongoDB server version in this repository (e.g. EOL'ing an old server
  version) and needing to find and remove now-dead version-gated tests, prose, and pseudocode across the whole

>   specifications repo.

Reviewer: ```suggestion
  specifications repo.
disable-model-invocation: true
```

Suggest making it an explicitly envoked skill to reduce context costs.

https://code.claude.com/docs/en/features-overview#context-cost-by-feature
> By default, skill descriptions load at session start so Claude can decide when to use them. Set `disable-model-invocation: true` in a skill’s frontmatter to hide it from Claude entirely until you invoke it manually. **_This reduces context cost to zero for skills you only trigger yourself._** For a skill you didn’t write, set [skillOverrides](https://code.claude.com/docs/en/skills#override-skill-visibility-from-settings) in settings to do the same without editing its file.

Author: I worry that the skill will go unnoticed and unused as a result.  How to weigh the tradeoff?

Reviewer 2: How much is the actual cost of the description being loaded here? The three lines of description text shouldn't add anything beyond a rounding error to the total context.

Author: The cost is low, and there aren't many skills in this repo.  I'm going to decline this one, as I think the tradeoff of someone not noticing the skill and being told of it outweighs the extra context.

---

# Retiring Server Versions

## Overview

Raising the minimum-supported-server-version floor doesn't just delete a few "if version < X" branches — it leaves
behind scattered dead conditionals, vacuous qualifiers, unrunnable tests, and stale prose across the whole repository.
The single biggest failure mode is scoping the cleanup too narrowly (to files already touched, or to obvious `< X.Y`
patterns) and then discovering — repeatedly, across multiple review rounds — that reviewers keep finding whole files and
whole *categories* of miss that the first pass never looked for.

**Core principle:** Establish fix/leave criteria explicitly before editing, sweep the *whole* repo (not just
already-changed files) with a wide pattern net, verify judgment calls against ground truth (real driver source), and
expect to re-sweep every time a reviewer finds a new *shape* of miss — not just patch the one instance.

## When to Use

- The project's minimum supported server version has just been raised (e.g. `DRIVERS-XXXX` "Remove pre-N.N references

>     from specs and tests").

Reviewer 3: Suggest noting this is only expected once all drivers have dropped support. That may help avoid losing test coverage in drivers that track specifications with a git submodule and have not-yet dropped the server version.

```suggestion
- When working on a ticket to remove old server references (e.g. `DRIVERS-XXXX` "Remove pre-N.N references from specs
    and tests").
    - Only expected once all drivers have dropped the old server (e.g. `DRIVERS-XXXX` "Mark Server version N.N as EOL" is
        complete).
```

Author: Applied locally

try to exclude that with lookarounds; it's a feature, not noise, since a patch-specific dead reference (e.g. a bug that
only affected `4.2.0`-`4.2.5`) is exactly the kind of hit you need to see and classify, not filter out. Multi-digit
collisions like `14.2` are excluded by the *left*-anchored patterns (`<`, `pre-`, since the anchor must sit immediately
before the digit) but NOT by the *right*-anchored ones (`+`, `or higher` — e.g. `14.2+` and `14.2 or higher` both still
match, since nothing constrains what precedes the number). This hasn't mattered yet (no MongoDB version starts with a
leading digit that creates a collision), but don't assume all four patterns are equally guarded — they aren't.

### Numeric comparisons

This is one single-floor pass (see Step 1 — don't bundle separate floor-raises together), and you already know both
numbers involved: the old floor being retired (e.g. 4.2) and the new floor (e.g. 4.4). Don't search broadly across every
historical version; that mostly re-surfaces content from *previous* retirements that was already correctly classified as

> historical narrative last time (tested: a broad `< ?[0-9]+\.[0-9]+` search returns 19 hits here, of which 10 are noise —

Reviewer 3: Consider removing sentences that reference exact hits for searches (e.g. "19 hits") as I expect this may become quickly stale, and I would expect does not have much use.

Author: Applied throughout.

done
```

Both commands print *every* `maxServerVersion` line, not just dead ones (tested in this repo: 22 raw file-level hits,
140 raw per-test hits) — including ones at or above the new floor (e.g. `maxServerVersion: "7.0.99"`), which are never
in scope (see Step 2 LEAVE: at/above the floor is fine). Read each hit's value against the new floor to classify it;
only after that filtering did this repo yield 7 real file-level hits (e.g.
`crud/tests/unified/deleteOne-hint-serverError.yml` with a top-level `maxServerVersion: 4.3.3`) and 53 real per-test
hits (e.g. `insertOne-serverErrors.yml`, where only 3 of its 8 tests are individually gated dead).

Unlike numbered markdown prose tests (Step 6 — mark `**Removed**`, never delete-and-renumber), these unified-test
entries are matched by description string, not by ordinal position, and aren't referenced by index across drivers — so
it's safe to delete them outright, no marker needed. Before deleting from a `.yml`, check whether it has a `.template`
source (see "Generated + `.template` file pairs" below) — if so, edit the template and regenerate via `make -C source`

> rather than hand-deleting the generated file directly.

Reviewer 3: `make -C source` does not process the `.template` files. Suggest describing the regeneration step in the referenced "Generated" section.

```suggestion
source (see "Generated + `.template` file pairs" below) — if so, edit the template and regenerate rather than hand-deleting the generated file directly.
```

Author: Applied locally

— don't assume a specific URL pattern will resolve directly (the release-notes site layout has changed and can
redirect); search for the page by title and follow the result. Each page enumerates exactly what was removed/changed in
that release. Confirm whether each removal landed *before* the new floor (in scope) or *at*/*after* it (out of scope,
same as anything at/above the floor), then grep the repo for the specific names you found. A "nothing removed at this
version is referenced in the repo" result is a legitimate, useful finding — don't treat an empty result as a failed
search.

### Generated + `.template` file pairs

Some YAML test files are generated, not hand-authored — check for a "Tests in this file are generated from
X.yml.template" comment at the top of the `.yml`, or a sibling file under `etc/templates/X.yml.template`. If you find
one, edit the `.template` source, not the generated `.yml` directly (a hand-edit to the generated file alone gets
silently overwritten the next time someone regenerates it), then run `make -C source` to propagate the change to the

> generated `.yml` and its `.json`. If `make -C source` fails, fix the tooling (e.g. a Node/js-yaml version mismatch) —

Reviewer 3: The `Makefile` does not process the `.template` files. Suggest referencing the GitHub workflows, which includes [steps to regenerate](https://github.com/mongodb/specifications/blob/0d6fee497b25bcb3617b8dc4f74e937f46a9cc99/.github/workflows/unified-tests.yml#L54-L61).
```suggestion
silently overwritten the next time someone regenerates it), then refer to `.github/workflows` to propagate the change to
the generated `.yml` and its `.json`. If regeneration fails, fix the tooling (e.g. a Node/js-yaml version mismatch) —
```

Author: Applied locally
